File: lab9/sneakers.py
from shoes import Shoes

class Sneakers(Shoes):
    """
    Клас Sneakers, що представляє кросівки. Наслідує клас Shoes та реалізує інтерфейс ShoeOperations.
    """

    def __init__(self, brand, size, color):
        """
        Конструктор для створення кросівок.

        :param brand: Бренд кросівок.
        :param size: Розмір кросівок.
        :param color: Колір кросівок.
        """
        super().__init__(brand, size, color)

    def wear(self):
        """
        Одягає кросівки, якщо вони не одягнуті.
        """
        if self.is_worn():
            print("Кросівки вже одягнені.")
        else:
            self.set_worn(True)
            print("Кросівки одягнені.")
            self.log_operation("Кросівки одягнені.")

    def remove(self):
        """
        Знімає кросівки, якщо вони надягнуті.
        """
        if not self.is_worn():
            print("Кросівки вже зняті.")
        else:
            self.set_worn(False)
            print("Кросівки зняті.")
            self.log_operation("Кросівки зняті.")

    def clean(self):
        """
        Очищає кросівки.
        """
        print("Кросівки очищені.")
        self.log_operation("Кросівки очищені.")

    def repair(self):
        """
        Відремонтувати кросівки, якщо вони зняті.
        """
        if self.is_worn():
            print("Спочатку зніміть кросівки, щоб відремонтувати їх.")
        else:
            print("Кросівки відремонтовані.")
            self.log_operation("Кросівки відремонтовані.")
            self.add_distance(-100)  # Скидаємо пробіг на 100 км після ремонту

    def polish(self):
        """
        Відполірувати кросівки, якщо вони зняті.
        """
        if self.is_worn():
            print("Спочатку зніміть кросівки, щоб відполірувати їх.")
        else:
            print("Кросівки відполіровані.")
            self.log_operation("Кросівки відполіровані.")

    def lace_up(self):
        """
        Зав'язує шнурки, якщо кросівки надягнуті.
        """
        if not self.is_worn():
            print("Спочатку одягніть кросівки, щоб зав'язати шнурки.")
        elif self.is_laced():
            print("Шнурки вже зав'язані.")
        else:
            self.set_laced(True)
            print("Шнурки зав'язані.")
            self.log_operation("Шнурки зав'язані.")

    def unlace(self):
        """
        Розв'язує шнурки, якщо вони зав'язані.
        """
        if not self.is_laced():
            print("Шнурки вже розв'язані.")
        else:
            self.set_laced(False)
            print("Шнурки розв'язані.")
            self.log_operation("Шнурки розв'язані.")

    def check_status(self):
        """
        Перевіряє та виводить поточний стан кросівок.
        """
        status = (
            f"Бренд: {self.get_brand()}, Розмір: {self.get_size()}, Колір: {self.get_color()}, "
            f"Шнурки зав'язані: {'так' if self.is_laced() else 'ні'}, "
            f"Одягнені: {'так' if self.is_worn() else 'ні'}, "
            f"Пробіг: {self.get_distance_run()} км"
        )
        if self.needs_repair():
            status += " (Потребує ремонту)"
        print(status)
        self.log_operation(f"Перевірено стан: {status}")

    def run(self, distance):
        """
        Виконує пробіг на задану відстань, якщо кросівки в належному стані.

        :param distance: Відстань для пробігу в км.
        """
        if self.needs_repair():
            print("Кросівки потребують ремонту. Спочатку відремонтуйте їх.")
            return

        if not self.is_laced():
            print("Потрібно зав'язати шнурки перед пробігом.")
            return

        self.add_distance(distance)
        print(f"Ви пробігли {distance} км.")
        self.log_operation(f"Пробігли {distance} км.")
        if self.needs_repair():
            print("Кросівки потребують ремонту після пробігу 100 км.")

    def log_operation(self, operation):
        """
        Записує проведену операцію в log-файл.
        """
        # Тут буде код для запису в log-файл.
        pass

File: Lab3/src/Ki_306/Petruk/Lab3/Sneakers.java

package Ki_306.Petruk.Lab3;

/**
 * Клас Sneakers, що представляє кросівки. Наслідує клас Shoes та реалізує інтерфейс ShoeOperations.
 */
public class Sneakers extends Shoes implements ShoeOperations {

    /**
     * Конструктор для створення кросівок.
     *
     * @param brand Бренд кросівок.
     * @param size Розмір кросівок.
     * @param color Колір кросівок.
     */
    public Sneakers(String brand, double size, String color) {
        super(brand, size, color);
    }

    /** Одягає кросівки, якщо вони не одягнуті. */
    @Override
    public void wear() {
        if (isWorn()) {
            System.out.println("Кросівки вже одягнені.");
        } else {
            setWorn(true);
            System.out.println("Кросівки одягнені.");
            logOperation("Кросівки одягнені.");
        }
    }

    /** Знімає кросівки, якщо вони надягнуті. */
    @Override
    public void remove() {
        if (!isWorn()) {
            System.out.println("Кросівки вже зняті.");
        } else {
            setWorn(false);
            System.out.println("Кросівки зняті.");
            logOperation("Кросівки зняті.");
        }
    }

    /** Очищає кросівки. */
    @Override
    public void clean() {
        System.out.println("Кросівки очищені.");
        logOperation("Кросівки очищені.");
    }

    /** Відремонтувати кросівки, якщо вони зняті. */
    @Override
    public void repair() {
        if (isWorn()) {
            System.out.println("Спочатку зніміть кросівки, щоб відремонтувати їх.");
        } else {
            System.out.println("Кросівки відремонтовані.");
            logOperation("Кросівки відремонтовані.");
            addDistance(-100); // Скидаємо пробіг на 100 км після ремонту
        }
    }

    /** Відполірувати кросівки, якщо вони зняті. */
    @Override
    public void polish() {
        if (isWorn()) {
            System.out.println("Спочатку зніміть кросівки, щоб відполірувати їх.");
        } else {
            System.out.println("Кросівки відполіровані.");
            logOperation("Кросівки відполіровані.");
        }
    }

    /** Зав'язує шнурки, якщо кросівки надягнуті. */
    @Override
    public void laceUp() {
        if (!isWorn()) {
            System.out.println("Спочатку одягніть кросівки, щоб зав'язати шнурки.");
        } else if (isLaced()) {
            System.out.println("Шнурки вже зав'язані.");
        } else {
            setLaced(true);
            System.out.println("Шнурки зав'язані.");
            logOperation("Шнурки зав'язані.");
        }
    }

    /** Розв'язує шнурки, якщо вони зав'язані. */
    @Override
    public void unlace() {
        if (!isLaced()) {
            System.out.println("Шнурки вже розв'язані.");
        } else {
            setLaced(false);
            System.out.println("Шнурки розв'язані.");
            logOperation("Шнурки розв'язані.");
        }
    }

    /**
     * Перевіряє та виводить поточний стан кросівок.
     */
    @Override
    public void checkStatus() {
        String status = "Бренд: " + getBrand() + ", Розмір: " + getSize() + ", Колір: " + getColor() +
                ", Шнурки зав'язані: " + (isLaced() ? "так" : "ні") +
                ", Одягнені: " + (isWorn() ? "так" : "ні") +
                ", Пробіг: " + getDistanceRun() + " км";
        if (needsRepair()) {
            status += " (Потребує ремонту)";
        }
        System.out.println(status);
        logOperation("Перевірено стан: " + status);
    }

    /**
     * Виконує пробіг на задану відстань, якщо кросівки в належному стані.
     *
     * @param distance Відстань для пробігу в км.
     */
    public void run(double distance) {
        if (needsRepair()) {
            System.out.println("Кросівки потребують ремонту. Спочатку відремонтуйте їх, щоб знову бігати.");
            return;
        }

        if (!isLaced()) {
            System.out.println("Потрібно зав'язати шнурки перед пробігом.");
            return;
        }

        addDistance(distance);
        System.out.println("Ви пробігли " + distance + " км.");
        logOperation("Пробігли " + distance + " км.");
        if (needsRepair()) {
            System.out.println("Кросівки потребують ремонту після пробігу 100 км.");
        }
    }

    /**
     * Записує проведену операцію в log-файл.
     *
     * @param operation Опис операції для запису в log-файл.
     */
    private void logOperation(String operation) {
        // Тут буде код для запису в log-файл.
    }
}
